perception/map_based_prediction/src/map_based_prediction_node.cpp

```diff
@@ -0,0 +1,997 @@
+  if (lanelet.second.centerline().size() <= 1) {
+    return false;
+  }
+
+  // Step2. Check if the obstacle is inside of this lanelet
+  if (!lanelet::geometry::inside(lanelet.second, search_point)) {
+    return false;
+  }
+
+  // If the object is in the objects history, we check if the target lanelet is
+  // inside the current lanelets id or following lanelets
+  const std::string object_id = toHexString(object.object_id);
+  if (objects_history_.count(object_id) != 0) {
+    const std::vector<lanelet::ConstLanelet> & possible_lanelet =
+      objects_history_.at(object_id).back().future_possible_lanelets;
+
+    bool not_in_possible_lanelet =
+      std::find(possible_lanelet.begin(), possible_lanelet.end(), lanelet.second) ==
+      possible_lanelet.end();
+    if (!possible_lanelet.empty() && not_in_possible_lanelet) {
+      return false;
+    }
+  }
+
+  // Step3. Calculate the angle difference between the lane angle and obstacle angle
+  double object_yaw = getObjectYaw(object);
+  const double lane_yaw = lanelet::utils::getLaneletAngle(
+    lanelet.second, object.kinematics.pose_with_covariance.pose.position);
+  const double delta_yaw = object_yaw - lane_yaw;
+  const double normalized_delta_yaw = std::atan2(std::sin(delta_yaw), std::cos(delta_yaw));
+  const double abs_norm_delta = std::fabs(normalized_delta_yaw);
+
+  // Step4. Check if the closest lanelet is valid, and add all
+  // of the lanelets that are below max_dist and max_delta_yaw
+  if (
+    lanelet.first < dist_threshold_for_searching_lanelet_ &&
+    abs_norm_delta < delta_yaw_threshold_for_searching_lanelet_) {
+    return true;
+  }
+
+  return false;
+}
+
+float MapBasedPredictionNode::calculateLocalLikelihood(
+  const lanelet::Lanelet & current_lanelet, const TrackedObject & object) const
+{
+  const auto & obj_point = object.kinematics.pose_with_covariance.pose.position;
+
+  // compute yaw difference between the object and lane
+  const double obj_yaw = getObjectYaw(object);
+  const double lane_yaw = lanelet::utils::getLaneletAngle(current_lanelet, obj_point);
+  const double delta_yaw = obj_yaw - lane_yaw;
+  const double abs_norm_delta_yaw = std::fabs(tier4_autoware_utils::normalizeRadian(delta_yaw));
+
+  // compute lateral distance
+  const auto centerline = current_lanelet.centerline();
+  std::vector<geometry_msgs::msg::Point> converted_centerline;
+  for (const auto & p : centerline) {
+    const auto converted_p = lanelet::utils::conversion::toGeomMsgPt(p);
+    converted_centerline.push_back(converted_p);
+  }
+  const double lat_dist =
+    std::fabs(tier4_autoware_utils::calcLateralOffset(converted_centerline, obj_point));
+
+  // Compute Chi-squared distributed (Equation (8) in the paper)
+  const double sigma_d = sigma_lateral_offset_;  // Standard Deviation for lateral position
+  const double sigma_yaw = M_PI * sigma_yaw_angle_ / 180.0;  // Standard Deviation for yaw angle
```

Comment:
change unit to [rad]. If [deg] is really needed, use `**_deg` suffix.

Reply:
done